spec/formatting.md

```diff
@@ -1,24 +1,124 @@
 
-The resolved value of a _text_ or a _literal_ is the character sequence of the _text_
-or _literal_ after any character escape has been converted to the escaped character.
-When a _literal_ is used as an _expression_ argument or on the right-hand side of an _option_,
-the formatting function MUST treat their resolved values the same independently of whether the
-value was originally _quoted_ or _unquoted_.
-For example, the _option_ `foo=42` and the _option_ `foo=|42|` 
-are treated as identical.
+- **_Resolution_** determines the value of a part of the message,
+  such as an _expression_, with reference to the current _formatting context_.
+  This can include multiple steps,
+  such as looking up the value of a variable and calling formatting functions.
+  The resolved value is not necessarily in the shape it will finally take,
+  but is "formattable", i.e. it contains everything required by the eventual formatting.
 
-## Variable Resolution
+  > For example, the resolved value of the _expression_ `{|0.40| :number style=percent}`
+  > could be an object such as
+  >
+  > ```
+  > { value: Number('0.40'),
+  >   formatter: NumberFormat(locale, { style: 'percent' }) }
+  > ```
+  >
+  > The shape of this value is implementation-dependent,
+  > and different implementations may choose to perform different levels of resolution.
+
+- **_Pattern Selection_** determines which of a message's _patterns_ is formatted.
+  For a message with no _selectors_, this is simple as there is only one _pattern_.
+  With _selectors_, this will depend on their _resolution_.
+
+- **_Formatting_** takes the resolved values of the selected _pattern_,
+  and formats them in the desired shape.
+  This specification only defines formatting messages as a single concatenated string,
+  but implementations SHOULD provide formatters for additional shapes
+  as appropriate for their setting.
+
+  > For example, a formatter in a web browser could format a message as a DOM fragment
+  > rather than as a representation of its HTML source.
+
+Formatter implementations are not required to expose
+the _resolution_ and _pattern selection_ operations to their users,
+or even use them in their internal processing,
+as long as the final _formatting_ result is made available to users
+and the observable behavior of the formatter matches that described here.
+
+## Formatting Context
+
+A message's **_formatting context_** represents the data and procedures that are required
+for the message's _resolution_, _pattern selection_ and _formatting_.
+
+At a minimum, it includes:
+
+- Information on the current **_locale_**,
+  potentially including a fallback chain of locales.
+  This will be passed on to formatting functions.
+
+- Information on the base directionality of the message and its _text_ tokens.
+  This will be used by strategies for bidirectional isolation.
+
+- A mapping of string identifiers to values,
+  defining variable values that may be used during _variable resolution_.
```

Comment:
Is it worth saying here that this mapping is for "external variables", as distinct from the variables defined with `let`-declarations?

Reply:
I'd prefer not adding to that here atm, as variable resolution is still being iterated upon.